pkgs/stdenv/generic/make-derivation.nix

```diff
@@ -441,6 +450,15 @@ else let
                    else true);
     };
 
+    checkedEnv = let
+      envNames = lib.attrNames env;
+      drvNames = lib.attrNames derivationArg;
+    in
+      assert lib.assertMsg (lib.mutuallyExclusive envNames drvNames) "The ‘env’ attribute set cannot contain any attributes passed to derivation. The following attributes are overlapping: ${lib.concatStringsSep ", " (lib.intersectLists envNames drvNames)}";
+      lib.mapAttrs
+        (n: v: assert lib.assertMsg (lib.isString v || lib.isBool v || lib.isInt v) "The ‘env’ attribute set can only contain string, boolean or integer attributes. The ‘${n}’ attribute is of type ${builtins.typeOf v}."; v)
+        env;
+
```

Comment:
These checks are symptomatic of the underlying problem which is that `drvAttrs` are still also environment variables.
The "top level" arguments of `mkDerivation` shouldn't have combined its roles as both a record of known parameters and a dictionary of environment variables and a set of exported attributes, as this combinations creates uncertainty around the intent of these attributes.

Reply:
Resolving because of scope creep, see https://github.com/NixOS/nixpkgs/pull/175649#discussion_r945235090